rustls/src/quic.rs

```diff
@@ -108,7 +145,62 @@ impl PacketKey {
                 &[],
             )),
             iv: hkdf_expand(secret, IvLen, b"quic iv", &[]),
+            suite,
+        }
+    }
+
+    /// Encrypt a QUIC packet
+    ///
+    /// `buf` contains the header, payload and space for the AEAD tag, with the payload starting at
+    /// `header_len` and ending at `tag_len()` before the end of `buf`.
+    pub fn encrypt(&self, packet: u64, buf: &mut [u8], header_len: usize) -> Result<(), Error> {
+        let (header, payload) = buf.split_at_mut(header_len);
+        let (payload, tag_storage) =
+            payload.split_at_mut(payload.len() - self.key.algorithm().tag_len());
+        let aad = aead::Aad::from(header);
+        let nonce = self.iv.nonce_for(packet);
+        let tag = self
+            .key
+            .seal_in_place_separate_tag(nonce, aad, payload)
+            .map_err(|_| Error::EncryptError)?;
+        tag_storage.copy_from_slice(tag.as_ref());
+        Ok(())
+    }
+
+    /// Decrypt a QUIC packet
+    ///
+    /// Returns the length of the decrypted payload.
+    pub fn decrypt(&self, packet: u64, header: &[u8], payload: &mut [u8]) -> Result<usize, Error> {
+        if payload.len() < self.key.algorithm().tag_len() {
+            return Err(Error::DecryptError);
         }
+
+        let payload_len = payload.len();
+        let aad = aead::Aad::from(header);
+        let nonce = self.iv.nonce_for(packet);
+        self.key
+            .open_in_place(nonce, aad, payload)
+            .map_err(|_| Error::DecryptError)?;
+        Ok(payload_len - self.key.algorithm().tag_len())
+    }
+
+    /// Number of times the packet key can be used without sacrificing confidentiality
+    ///
+    /// See https://www.rfc-editor.org/rfc/rfc9001.html#name-confidentiality-limit.
+    pub fn confidentiality_limit(&self) -> u64 {
+        self.suite.confidentiality_limit
+    }
+
+    /// Number of times the packet key can be used without sacrificing integrity
+    ///
+    /// https://www.rfc-editor.org/rfc/rfc9001.html#name-integrity-limit.
+    pub fn integrity_limit(&self) -> u64 {
+        self.suite.integrity_limit
+    }
+
+    /// Tag length for the underlying AEAD algorithm
+    pub fn tag_len(&self) -> usize {
```

Comment:
First, if we need this it probably should be `#[inline]`.

But, do we really need this? I would hope that this kind of detail would be encapsulated by the `encrypt` and `decrypt` functions.

Reply:
The QUIC implementation needs to know how much space to leave at the end of the packet, so that the tag can be written without exceeding the MTU.

Reply:
Alternatively we could return an `impl AsRef<[u8]>` here or something, akin to the separate tag approach that *ring* uses.

Reply:
(It's `#[inline]` now.)

Reply:
This is resolved from my end.